### ISSUE TITLE
Request for standalone visualization script (Figure 5)

### ISSUE THREAD

**Issue by chenyao2005** [GrokCV/GrokCSO#8]
Hi authors,

First of all, thank you for sharing your excellent work and open-sourcing the code! I am a prospective graduate student, and your paper has been incredibly helpful for my learning in infrared small target detection.

I have successfully reproduced the training and testing processes for DISTA-Net on my local environment (RTX 3080 Ti). I am very happy to share that the CSO-mAP results for c=3, c=5, and c=7 align perfectly with the results reported in Table 1 and Table 3 of your paper. (I also successfully evaluated the PSNR/SSIM metrics after referencing your helpful reply in closed Issue #5!)

There is only one final piece missing in my reproduction that I would like to seek your help with:

Standalone Visualization Script (like Figure 5)
I tried to visualize the unmixing results using the --show-dir argument in test.py, but it caused an AttributeError related to DetVisualizationHook (as it seems the framework natively expects bounding box formats rather than the specific unmixing arrays).

May I ask if you have a standalone Python or MATLAB script used to generate the beautiful unmixing comparison images shown in Figure 5? It would be wonderful to visually inspect the results of the models I just trained.

Thank you again for your time and contribution to the community. Looking forward to your reply!

Best regards,
Ariel

**Comment by Oak777** [comment]
Hi Ariel,

Thank you for your kind words and interest in our work! We are glad to hear that you have successfully reproduced the results for DISTA-Net.

Regarding the visualization in Figure 5, you are correct that the standard DetVisualizationHook in MMEngine is designed for bounding boxes and does not directly support the sub-pixel unmixing maps used in our paper.

To generate similar visualizations, we typically extract a specific ROI (Region of Interest) where targets are concentrated and render them as sub-pixel grids. I have integrated the core visualization logic into the CSO_Metrics class below. This allows you to generate individual prediction images (like the small tiles in Figure 5) directly during the inference process by setting vis=True.


#### Updated CSO_Metrics with Visualization Support

```python
from collections import OrderedDict
from typing import List, Optional
import numpy as np
import os
import torch
import matplotlib.pyplot as plt
import matplotlib.patches as patches
from mmengine.evaluator import BaseMetric
from mmengine.logging import MMLogger
from mmengine.registry import METRICS
from ..functional import eval_map2

@METRICS.register_module()
class CSO_Metrics(BaseMetric):
    """
    CSO_Metrics class with integrated visualization for sub-pixel reconstruction.
    """
    default_prefix: Optional[str] = 'cso_metric'

    def __init__(self,
                 distance_thrs=(0.05, 0.1, 0.15, 0.2, 0.25),
                 brightness_threshold: float = 50,
                 c=3,
                 collect_device: str = 'cpu',
                 prefix: Optional[str] = None,
                 vis: bool = False,
                 vis_dir: str = 'vis_results') -> None:
        super().__init__(collect_device=collect_device, prefix=prefix)
        self.distance_thrs = list(distance_thrs)
        self.brightness_threshold = brightness_threshold
        self.c = c
        self.vis = vis
        self.vis_dir = vis_dir
        
        if self.vis and not os.path.exists(self.vis_dir):
            os.makedirs(self.vis_dir, exist_ok=True)

    def get_best_crop_origin(self, objects, full_w=11, crop_size=3):
        """
        Calculate the top-left corner of the best ROI based on predicted points.
        """
        if not objects:
            return 0, 0
        xs = [obj['xc'] for obj in objects]
        ys = [obj['yc'] for obj in objects]
        
        # Center the crop around the mean position of targets
        start_x = int(np.mean(xs) - 0.5)
        start_y = int(np.mean(ys) - 0.5)
        
        # Clamp to image boundaries
        max_idx = full_w - crop_size
        start_x = max(0, min(start_x, max_idx))
        start_y = max(0, min(start_y, max_idx))
        return start_x, start_y

    def visualize_sr_result(self, objects, start_x, start_y, save_path):
        """
        Renders the ROI as a discrete grid on a black background.
        """
        plt.switch_backend('Agg')
        scale = self.c
        crop_size = self.c
        limit = crop_size * scale
        
        fig, ax = plt.subplots(figsize=(6, 6))
        fig.patch.set_facecolor('black')
        ax.set_facecolor('black')

        for obj in objects:
            # Convert global coordinates to ROI-relative grid coordinates
            grid_x = int((obj['xc'] - start_x) * scale)
            grid_y = int((obj['yc'] - start_y) * scale)

            if 0 <= grid_x < limit and 0 <= grid_y < limit:
                val = min(max(obj['brightness'] / 255.0, 0.0), 1.0)
                # Use Rectangle to create the discrete pixel look
                rect = patches.Rectangle((grid_x, grid_y), 1, 1, 
                                         linewidth=0, facecolor=(val, val, val))
                ax.add_patch(rect)

        ax.set_xlim(0, limit)
        ax.set_ylim(limit, 0) # Flip Y to match image coordinates (row major)
        ax.axis('off')
        plt.savefig(save_path, dpi=100, bbox_inches='tight', pad_inches=0, facecolor='black')
        plt.close()

    def process(self, data_batch: dict, outputs) -> None:
        file_ids = data_batch.get("file_id", [None] * len(outputs))
        output = outputs[0]
        x_output = output['x_final'].cpu().numpy()
        ann_list = output['targets_GT']
        x_output[x_output < self.brightness_threshold] = 0

        for idx in range(x_output.shape[0]):
            ann = ann_list[idx]
            matrix = x_output[idx].reshape(11 * self.c, 11 * self.c)
            non_zero_indices = np.nonzero(matrix)
            
            dets = []
            pred_objects_for_vis = []
            
            for i in range(len(non_zero_indices[0])):
                row, col = non_zero_indices[0][i], non_zero_indices[1][i]
                val = matrix[row, col]
                
                # Map back to 11x11 coordinate space
                coord_row = float(1.0 * (row - (self.c - 1) // 2) / self.c)
                coord_col = float(1.0 * (col - (self.c - 1) // 2) / self.c)
                
                dets.append([coord_row, coord_col, val])
                pred_objects_for_vis.append({'xc': coord_row, 'yc': coord_col, 'brightness': val})

            self.results.append((ann, dets))

            # Trigger visualization
            if self.vis:
                sx, sy = self.get_best_crop_origin(pred_objects_for_vis, crop_size=self.c)
                save_path = os.path.join(self.vis_dir, f"{file_ids[idx]}_pred.png")
                self.visualize_sr_result(pred_objects_for_vis, sx, sy, save_path)

    def compute_metrics(self, results: list) -> dict:
        logger: MMLogger = MMLogger.get_current_instance()
        gts, preds = zip(*results)
        eval_results = OrderedDict()
        mean_aps = []

        for distance_thr in self.distance_thrs:
            mean_ap, _ = eval_map2(preds, gts, distance_thr=distance_thr, logger=logger)
            mean_aps.append(mean_ap)
            eval_results[f'AP{int(distance_thr * 100):02d}'] = round(mean_ap, 5)

        eval_results['mAP'] = sum(mean_aps) / len(mean_aps)
        eval_results.move_to_end('mAP', last=False)
        return eval_results
```

#### Important Implementation Details:

1.  **ROI Logic (`get_best_crop_origin`)**: This function determines the $c \times c$ grid coordinates (the physical pixel area) to crop. By centering it on the target predictions, you get the zoomed-in view seen in Figure 5.
2.  **Sub-pixel Grid Rendering (`visualize_sr_result`)**: We use matplotlib.patches.Rectangle to fill each corresponding sub-pixel grid with the predicted target radiation intensity.
3.  **Note on SSIM**: Regarding the similarity metrics mentioned in your issue, we actually **do not recommend using SSIM** for this task. It provides very little meaningful information for infrared small target detection; it was added only to satisfy a reviewer's request during the revision process. We consider it an oversight that it wasn't removed in the final Camera-Ready version.

We hope this clarifies the visualization process!

Best regards,  
DISTA-Net Authors

**Comment by chenyao2005** [comment]
Hi DISTA-Net Authors,

Wow, thank you so much for the incredibly detailed and helpful response! Providing the completely updated CSO_Metrics class with the ROI and grid rendering logic is beyond what I expected. I really appreciate your time.

Also, thank you for the candid clarification regarding the SSIM metric. It's actually a huge relief to hear that, as I had suspected it might not be very discriminative given the nature of the dark background in IRST tasks. Knowing it was just a reviewer's request clears up a lot of confusion!

I will integrate this visualization code into my pipeline right away. Thanks again for open-sourcing such a fantastic piece of work and for being so supportive of the community.

Best regards,
Ariel
